DESCRIPTION:-

Background of Problem Statement :

The purpose of the project is to use one or more of the classification algorithms to train a model on the phishing website dataset

You are provided with the following resources that can be used as inputs for your model:
1. A collection of website URLs for 11000+ wesites. Each sample has 30 website
parameters and a class label identifying it as a phishing website or not (1 or -1)
2. Code template containing these code blocks:
a. Import modules (Part 1)
b. Load data function + input/ouput field descriptions

You are expected to write the code for a binary classification model (phishing website or not) using python scikit - learn that trains on the data and calculates the accuracy score on the test data.

The purpose of the project is to use one or more of the classification algorithms to train a model on the phishing website dataset

You are provided with the following resources that can be used as inputs for your model:
1. A collection of website URLs for 11000+ wesites. Each sample has 30 website parameters and a class label identifying it as a phishing website or not (1 or -1)
2. Code template containing these code blocks:
a. Import modules (Part 1)
b. Load data function + input/ouput field descriptions

You are expected to write the code for a binary classification model (phishing website or not) using python scikit - learn that trains on the data and calculates the accuracy score on the test data.


Objective:

A phishing website is a common social engineering method that mimics trustful uniform resource locators (URLs) and webpages. 
The objective of this project is to train machine learning models and deep neural nets on the dataset created to predict phishing websites. 
Both phishing and benign URLs of websites are gathered to form a dataset and from them required URL and website content-based features are extracted. 
The performance level of each model is measures and compared.


Domain : Cyber security and Web Minning
    

Implementation Details :

-Load the data set 
Loaded the dataset into the program using package pandas and extracted input (X) & output (y) data from dataset.

-Split the features and label into training and testing data
Performed splitting of data set into training data (70%) and testing data (30%) using train_test_split.

-Standarized data
standarized data using standard scaler.

-Perform Logistics Regression Classifier
Implemented  logistic Regression Classifier on training data with "C" parameter = 100 an d printed the predicted the output.
Printed Confusion Matrix to specify the count of misclassified samples in the test data prediction.

-Extracted input (X : prefix_suffix and URL_of_Anchor) and output (y) data from the data set with index 5.

-Split the features and label into training and testing data
Performed splitting of data set into training data (70%) and testing data (30%) using train_test_split.

-Standarize data
Standardized data using standard scaler 

-Peform Logistic Regression Classifier 
Implemented Logistics Regression classifier on training data with "C" parameter = 100and printed the predicted the output.
Printed Confusion Matrix to specify the count of misclassified samples in the test data prediction 

- Visualize the outcome of Logistic Regression Classifier Test set.


In [1]:
#Importing Necessary Libraries:

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

from sklearn.preprocessing import StandardScaler 
from sklearn.linear_model  import  LogisticRegression
from sklearn.model_selection import train_test_split

from sklearn.metrics import accuracy_score, confusion_matrix
import seaborn as sns

sns.set()

In [2]:
#loading the Dataset

df_phishing = pd.read_csv(r"C:\Users\prash\phishing.txt")

In [3]:
df_phishing.shape

(11054, 31)

In [4]:
df_phishing.head()

-1  1  1.1  1.2  -1.1  -1.2  -1.3  -1.4  -1.5  1.3  ...  1.9  1.10  -1.11  \
0   1  1    1    1     1    -1     0     1    -1    1  ...    1     1     -1   
1   1  0    1    1     1    -1    -1    -1    -1    1  ...    1     1      1   
2   1  0    1    1     1    -1    -1    -1     1    1  ...    1     1     -1   
3   1  0   -1    1     1    -1     1     1    -1    1  ...   -1     1     -1   
4  -1  0   -1    1    -1    -1     1     1    -1    1  ...    1     1      1   

   -1.12  -1.13  -1.14  1.11  1.12  -1.15  -1.16  
0     -1      0     -1     1     1      1     -1  
1     -1      1     -1     1     0     -1     -1  
2     -1      1     -1     1    -1      1     -1  
3     -1      0     -1     1     1      1      1  
4      1      1     -1     1    -1     -1      1  

[5 rows x 31 columns]

In [5]:
df_phishing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11054 entries, 0 to 11053
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   -1      11054 non-null  int64
 1   1       11054 non-null  int64
 2   1.1     11054 non-null  int64
 3   1.2     11054 non-null  int64
 4   -1.1    11054 non-null  int64
 5   -1.2    11054 non-null  int64
 6   -1.3    11054 non-null  int64
 7   -1.4    11054 non-null  int64
 8   -1.5    11054 non-null  int64
 9   1.3     11054 non-null  int64
 10  1.4     11054 non-null  int64
 11  -1.6    11054 non-null  int64
 12  1.5     11054 non-null  int64
 13  -1.7    11054 non-null  int64
 14  1.6     11054 non-null  int64
 15  -1.8    11054 non-null  int64
 16  -1.9    11054 non-null  int64
 17  -1.10   11054 non-null  int64
 18  0       11054 non-null  int64
 19  1.7     11054 non-null  int64
 20  1.8     11054 non-null  int64
 21  1.9     11054 non-null  int64
 22  1.10    11054 non-null  int64
 23  -1.11   110

In [6]:
# Missing Values

df_phishing.isnull().sum()

-1       0
1        0
1.1      0
1.2      0
-1.1     0
-1.2     0
-1.3     0
-1.4     0
-1.5     0
1.3      0
1.4      0
-1.6     0
1.5      0
-1.7     0
1.6      0
-1.8     0
-1.9     0
-1.10    0
0        0
1.7      0
1.8      0
1.9      0
1.10     0
-1.11    0
-1.12    0
-1.13    0
-1.14    0
1.11     0
1.12     0
-1.15    0
-1.16    0
dtype: int64

In [7]:
df_phishing.describe()

-1             1           1.1           1.2          -1.1  \
count  11054.000000  11054.000000  11054.000000  11054.000000  11054.000000   
mean       0.313914     -0.633345      0.738737      0.700561      0.741632   
std        0.949495      0.765973      0.674024      0.713625      0.670837   
min       -1.000000     -1.000000     -1.000000     -1.000000     -1.000000   
25%       -1.000000     -1.000000      1.000000      1.000000      1.000000   
50%        1.000000     -1.000000      1.000000      1.000000      1.000000   
75%        1.000000     -1.000000      1.000000      1.000000      1.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

               -1.2          -1.3          -1.4          -1.5           1.3  \
count  11054.000000  11054.000000  11054.000000  11054.000000  11054.000000   
mean      -0.734938      0.064049      0.251040     -0.336711      0.628551   
std        0.678165      0.817492      0.911856      0.941651      0.777804   
min       -1.000000     -1.000000     -1.000000     -1.000000     -1.000000   
25%       -1.000000     -1.000000     -1.000000     -1.000000      1.000000   
50%       -1.000000      0.000000      1.000000     -1.000000      1.000000   
75%       -1.000000      1.000000      1.000000      1.000000      1.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

       ...           1.9          1.10         -1.11         -1.12  \
count  ...  11054.000000  11054.000000  11054.000000  11054.000000   
mean   ...      0.613353      0.816899      0.061335      0.377239   
std    ...      0.789845      0.576807      0.998162      0.926158   
min    ...     -1.000000     -1.000000     -1.000000     -1.000000   
25%    ...      1.000000      1.000000     -1.000000     -1.000000   
50%    ...      1.000000      1.000000      1.000000      1.000000   
75%    ...      1.000000      1.000000      1.000000      1.000000   
max    ...      1.000000      1.000000      1.000000      1.000000   

              -1.13         -1.14          1.11          1.12         -1.15  \
count  11054.000000  11054.000000  11054.000000  11054.000000  11054.000000   
mean       0.287407     -0.483626      0.721549      0.343948      0.719739   
std        0.827680      0.875314      0.692395      0.569936      0.694276   
min       -1.000000     -1.000000     -1.000000     -1.000000     -1.000000   
25%        0.000000     -1.000000      1.000000      0.000000      1.000000   
50%        1.000000     -1.000000      1.000000      0.000000      1.000000   
75%        1.000000      1.000000      1.000000      1.000000      1.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

              -1.16  
count  11054.000000  
mean       0.113986  
std        0.993527  
min       -1.000000  
25%       -1.000000  
50%        1.000000  
75%        1.000000  
max        1.000000  

[8 rows x 31 columns]

In [8]:
# Classify as features and label

X = df_phishing.iloc[:,:-1].values
y = df_phishing.iloc[:,30].values

In [9]:
#split features and label into training and testing data

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.3,random_state=4)

In [10]:
# Applying Logistic Regression

log_reg = LogisticRegression()

log_reg.fit(X_train,y_train)

LogisticRegression()

In [11]:
y_pred = log_reg.predict(X_test)

In [12]:
conf_mat = confusion_matrix(y_test,y_pred)
conf_mat

array([[1368,  142],
       [  96, 1711]], dtype=int64)

In [13]:
true_positive = conf_mat[0][0]
false_positive = conf_mat[0][1]
false_negative = conf_mat[1][0]
true_negative = conf_mat[1][1]

In [14]:
# Accuracy

Accuracy = (true_positive + true_negative) / (true_positive +false_positive + false_negative + true_negative)
Accuracy

0.928248417244498

In [15]:
# Precison

Precision = true_positive/(true_positive+false_positive)
Precision

0.9059602649006623

In [16]:
# Recall

Recall = true_positive/(true_positive+false_negative)
Recall

0.9344262295081968

In [17]:
# Fscore

f1=2*((Precision * Recall )) /(Precision + Recall)
f1

0.9199731002017485

In [18]:
# performing feature scaling

from sklearn.preprocessing import StandardScaler
scalar = StandardScaler()
X_train = scalar.fit_transform (X_train)
X_test = scalar.fit_transform (X_test) 

In [19]:
#Logistic Regression Classifier

from sklearn.linear_model import LogisticRegression
LRclassifier = LogisticRegression(C=100,random_state=0)
LRclassifier.fit(X_train,y_train)

LogisticRegression(C=100, random_state=0)

In [20]:
LRpredict = LRclassifier.predict(X_test)

In [21]:
#LRC training score
LRclassifier.score(X_train,y_train)

0.9298177588212485

In [22]:
#LRC test score
LRclassifier.score(X_test,y_test)

0.9267410310521556

Evaluating By Various Machine Learning Models

From the dataset above, it is clear that this is a supervised machine learning task. There are two major types of supervised machine learning problems, called classification and regression.

This data set comes under classification problem, as the input URL is classified as phishing (1) or legitimate (0). The supervised machine learning models (classification) considered to train the dataset in this notebook are:

Decision Tree
Random Forest
XGBoost
Autoencoder Neural Network
Support Vector Machines

In [23]:
# Creating holders to store the model performance results
ML_Model = []
acc_train = []
acc_test = []

#function to call for storing the results
def storeResults(model, a,b):
  ML_Model.append(model)
  acc_train.append(round(a, 3))
  acc_test.append(round(b, 3))

In [24]:
# Decision Tree model 
from sklearn.tree import DecisionTreeClassifier

# instantiate the model 
tree = DecisionTreeClassifier(max_depth = 5)
# fit the model 
tree.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=5)

In [25]:
#predicting the target value from the model for the samples
y_test_tree = tree.predict(X_test)
y_train_tree = tree.predict(X_train)

In [26]:
#computing the accuracy of the model performance
acc_train_tree = accuracy_score(y_train,y_train_tree)
acc_test_tree = accuracy_score(y_test,y_test_tree)

print("Decision Tree: Accuracy on training Data: {:.3f}".format(acc_train_tree))
print("Decision Tree: Accuracy on test Data: {:.3f}".format(acc_test_tree))

Decision Tree: Accuracy on training Data: 0.918
Decision Tree: Accuracy on test Data: 0.922


In [27]:
#storing the results. 

storeResults('Decision Tree', acc_train_tree, acc_test_tree)

In [28]:
# Random Forest model
from sklearn.ensemble import RandomForestClassifier

# instantiate the model
forest = RandomForestClassifier(max_depth=5)

# fit the model 
forest.fit(X_train, y_train)

RandomForestClassifier(max_depth=5)

In [29]:
#predicting the target value from the model for the samples
y_test_forest = forest.predict(X_test)
y_train_forest = forest.predict(X_train)

In [30]:
#computing the accuracy of the model performance
acc_train_forest = accuracy_score(y_train,y_train_forest)
acc_test_forest = accuracy_score(y_test,y_test_forest)

print("Random forest: Accuracy on training Data: {:.3f}".format(acc_train_forest))
print("Random forest: Accuracy on test Data: {:.3f}".format(acc_test_forest))

Random forest: Accuracy on training Data: 0.933
Random forest: Accuracy on test Data: 0.938


In [31]:
#storing the results.
storeResults('Random Forest', acc_train_forest, acc_test_forest)

In [32]:
#XGBoost Classification model
from xgboost import XGBClassifier

# instantiate the model
xgb = XGBClassifier(learning_rate=0.4,max_depth=7)
#fit the model
xgb.fit(X_train, y_train)

[23:29:02] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


C:\Users\prash\anaconda3\envs\py38\lib\site-packages\xgboost\sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.4, max_delta_step=0, max_depth=7,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=8, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [33]:
#predicting the target value from the model for the samples
y_test_xgb = xgb.predict(X_test)
y_train_xgb = xgb.predict(X_train)

In [34]:
#computing the accuracy of the model performance
acc_train_xgb = accuracy_score(y_train,y_train_xgb)
acc_test_xgb = accuracy_score(y_test,y_test_xgb)

print("XGBoost: Accuracy on training Data: {:.3f}".format(acc_train_xgb))
print("XGBoost : Accuracy on test Data: {:.3f}".format(acc_test_xgb))

XGBoost: Accuracy on training Data: 0.990
XGBoost : Accuracy on test Data: 0.970


In [35]:
#storing the results

storeResults('XGBoost', acc_train_xgb, acc_test_xgb)

In [36]:
#importing required packages
import keras
from keras.layers import Input, Dense
from keras import regularizers
import tensorflow as tf
from keras.models import Model
from sklearn import metrics

In [37]:
#building autoencoder model

input_dim = X_train.shape[1]
encoding_dim = input_dim

input_layer = Input(shape=(input_dim, ))
encoder = Dense(encoding_dim, activation="relu",
                activity_regularizer=regularizers.l1(10e-4))(input_layer)
encoder = Dense(int(encoding_dim), activation="relu")(encoder)

encoder = Dense(int(encoding_dim+2), activation="relu")(encoder)
code = Dense(int(encoding_dim+4), activation='relu')(encoder)
decoder = Dense(int(encoding_dim+2), activation='relu')(code)

decoder = Dense(int(encoding_dim), activation='relu')(encoder)
decoder = Dense(input_dim, activation='relu')(decoder)
autoencoder = Model(inputs=input_layer, outputs=decoder)
autoencoder.summary()

Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 30)]              0         
_________________________________________________________________
dense (Dense)                (None, 30)                930       
_________________________________________________________________
dense_1 (Dense)              (None, 30)                930       
_________________________________________________________________
dense_2 (Dense)              (None, 32)                992       
_________________________________________________________________
dense_5 (Dense)              (None, 30)                990       
_________________________________________________________________
dense_6 (Dense)              (None, 30)                930       
Total params: 4,772
Trainable params: 4,772
Non-trainable params: 0
___________________________________________________________

In [38]:
#compiling the model
autoencoder.compile(optimizer='adam',
                    loss='binary_crossentropy',
                    metrics=['accuracy'])

#Training the model
history = autoencoder.fit(X_train, X_train, epochs=10, batch_size=64, shuffle=True, validation_split=0.2)

Epoch 1/10
97/97 [==============================] - 4s 31ms/step - loss: -0.1418 - accuracy: 0.0107 - val_loss: -1.3289 - val_accuracy: 0.0181
Epoch 2/10
97/97 [==============================] - 0s 4ms/step - loss: -1.2109 - accuracy: 0.0238 - val_loss: -1.8881 - val_accuracy: 0.0433
Epoch 3/10
97/97 [==============================] - 0s 4ms/step - loss: -1.7577 - accuracy: 0.0287 - val_loss: -1.9399 - val_accuracy: 0.0194
Epoch 4/10
97/97 [==============================] - 0s 4ms/step - loss: -1.8430 - accuracy: 0.0216 - val_loss: -2.0896 - val_accuracy: 0.0155
Epoch 5/10
97/97 [==============================] - 0s 4ms/step - loss: -2.0845 - accuracy: 0.0190 - val_loss: -2.3845 - val_accuracy: 0.0252
Epoch 6/10
97/97 [==============================] - 0s 5ms/step - loss: -2.1531 - accuracy: 0.0192 - val_loss: -2.3293 - val_accuracy: 0.0149
Epoch 7/10
97/97 [==============================] - 0s 4ms/step - loss: -2.1056 - accuracy: 0.0314 - val_loss: -2.4196 - val_accuracy: 0.0465
Epoch

In [39]:
acc_train_auto = autoencoder.evaluate(X_train, X_train)[1]
acc_test_auto = autoencoder.evaluate(X_test, X_test)[1]

print('\nAutoencoder: Accuracy on training Data: {:.3f}' .format(acc_train_auto))
print('Autoencoder: Accuracy on test Data: {:.3f}' .format(acc_test_auto))

104/104 [==============================] - 0s 2ms/step - loss: -2.5543 - accuracy: 0.0775

Autoencoder: Accuracy on training Data: 0.078
Autoencoder: Accuracy on test Data: 0.077


In [40]:
#storing the result.

storeResults('AutoEncoder', acc_train_auto, acc_test_auto)

In [41]:
#Support vector machine model
from sklearn.svm import SVC

# instantiate the model
svm = SVC(kernel='linear', C=1.0, random_state=12)
#fit the model
svm.fit(X_train, y_train)

SVC(kernel='linear', random_state=12)

In [42]:
#predicting the target value from the model for the samples
y_test_svm = svm.predict(X_test)
y_train_svm = svm.predict(X_train)

In [43]:
#computing the accuracy of the model performance
acc_train_svm = accuracy_score(y_train,y_train_svm)
acc_test_svm = accuracy_score(y_test,y_test_svm)

print("SVM: Accuracy on training Data: {:.3f}".format(acc_train_svm))
print("SVM : Accuracy on test Data: {:.3f}".format(acc_test_svm))

SVM: Accuracy on training Data: 0.929
SVM : Accuracy on test Data: 0.925


In [44]:
#storing the results.

storeResults('SVM', acc_train_svm, acc_test_svm)

Comparision of Models

To compare the models performance, a dataframe is created. The columns of this dataframe are the lists created to store the results of the model.

In [45]:
#creating dataframe
results = pd.DataFrame({ 'ML Model': ML_Model,    
    'Train Accuracy': acc_train,
    'Test Accuracy': acc_test})
results

ML Model  Train Accuracy  Test Accuracy
0  Decision Tree           0.918          0.922
1  Random Forest           0.933          0.938
2        XGBoost           0.990          0.970
3    AutoEncoder           0.078          0.077
4            SVM           0.929          0.925

In [46]:
#Sorting the datafram on accuracy
results.sort_values(by=['Test Accuracy', 'Train Accuracy'], ascending=False)

ML Model  Train Accuracy  Test Accuracy
2        XGBoost           0.990          0.970
1  Random Forest           0.933          0.938
4            SVM           0.929          0.925
0  Decision Tree           0.918          0.922
3    AutoEncoder           0.078          0.077

Conclusion : From the above comparision, it is clear that the XGBoost Classifier works well with this dataset.

The Accuracy of model using logistic Regression is 0.92

From all the models the Logistics regression works well with the data set with accuracy of 0.92